20240627-BC-listapi.md

```diff
@@ -0,0 +1,182 @@
+# List API for the Dapr state component
+
+This proposal proposes implementing a List API in Dapr's state component. The List API will enable the retrieval of keys in a state store based on certain criteria, providing users with the necessary visibility into stored keys. List API results will **not** include the value. SDKs could provide the possibility to bulk get all the keys returned in a page.
+
+The requirements for the API are:
+
+- Ability to list all keys in a state store
+- Ability to list keys in a state store with a certain prefix
+- The results can be sorted
+- The results can be paginated
+
+As with the other state store APIs, the List API will also have the difficult job of finding a set of features that are supported across most state store components and filling in the gaps with reasonable behaviour when they aren’t. 
+
+## API
+
+### HTTP
+
+Developers can list keys by issuing an HTTP API call to the Dapr sidecar:
+
+```bash
+GET /v1.0/state/:storeName/?prefix={prefix}&sorting={sorting}&page_limit={pageLimit}&page_token={pageToken}
+```
+
+The `sorting` query parameter can accept one of the following values:
+- `default`
+- `asc`
+- `desc`
+
+
+The response will be a JSON object with the following structure:
+```json
+{
+    "keys": ["key1", "key2", "key3", "...", "keyN"],
+    "next_page_token": "nextTokenString"
+}
+```
+
+For example:  
+Request:
+```cURL
+GET /v1.0/state/myStateStore?prefix=user&sorting=asc&page_limit=3&page_token=user3
+```
+Response:
+```json
+{
+    "keys": ["user4", "user5", "user6"],
+    "next_page_token": "user6"
+}
+```
+
+### gRPC
+
+Developers can also list keys by issuing a unary gRPC call
+
+```bash
+service Dapr {
+    ...
+	rpc ListState(ListRequest) returns (ListResponse) {}
+	...
+}
+
+message ListStateRequest {
+  // The prefix that should be used for listing.
+  optional string prefix = 1;
```

Comment:
I think prefix matching is important for apps to filter based on customer ID, scenarios like: all orders where key starts with "Customer1035143531|" since keys are composed as "customer Id|Order Id".

Reply:
How should case-sensitivity be handled, if at all?